proposals/0000-dps-compact.md

```diff
@@ -0,0 +1,393 @@
+
+parseWithDest (parse directly into compact region): OK
+  136  ms ± 5.4 ms, 283 MB allocated,  37 MB copied, 269 MB peak memory
+```
+
+(`***`_The  program only parses the document and then shutdowns, so in a real-world use case doing something non trivial with the parsed data, the time spent by the GC traversing the parsed dataset again and again in `parseWithoutDest (parse into GC heap, stay in GC heap` would be significantly more important._)
+
+Unlike the previous example, here the code for the parser is very similar and equally verbose for both the naive and DPS version, so there isn't much extra effort needed to get these performance benefits.
+
+## Effect and Interactions
+
+Primitive `compactAddHollow#` allow for direct allocation in a compact region, without copy, as stated in the [Motivation](#motivation) section. It needs to be combined with `reifyInfoTablePtr#` to properly allocate well-formed constructor objects with the right info table pointer. The pair is flexible enough to allow destination-passing style programming in Haskell, which allows for implementation of all examples presented above.
+
+I don't think the proposed changes will have any specific interaction with other GHC features. The implementation of the new primitives makes very little assumptions on the existing internals of GHC.
+
+reifyInfoTablePtr#` and `LiftedCtorToSymbol` are not harmful in any way; they reduces to a static value at compile time or raise a compile-time error.
+
+`compactAddHollow#` is a bit more dangerous, as it allows for creation of a constructor object in a compact region that should not be read without further modification, and that shouldn't be moved either (or that would invalidate the pointers that the user will use to write to its fields). In that sense, it is more unsafe than existing `compactAdd#` and `compactAddWithSharing#` primitives. That being said, the compact region guarantees wrt. the garbage collector should be enough to ensure defined behavior of a program using this new primitive.
+
+The library code makes many assumptions on the representation of Haskell heap objects, but it isn't in the perimeter of this proposal.
+
+## Costs and Drawbacks
+
+The changes in this proposal should have a very minimal impact on GHC maintenance. The changes are quite localized.
+
+## Backward Compatibility
+
+0. No breakage should be expected for existing code.
+
+## Alternatives
+
+The pair of primops `compactAddHollow#` and `reifyInfoTablePtr#` are made to work together, and I don't see any alternative design than a 2-stage process (compile-time and then runtime) to effectively allocate a chosen constructor object in a compact region without a copy. I didn't feel like I had to do many choices when designing these, but I'm still listening to feedback and alternative ideas.
+
+The rest of the proposal however has more room for alternatives. I listed the few that came to my mind:
```

Comment:
You don't need to answer to questions that don't exist. Simply list alternative designs and explains why you chose what you chose, or at least places where questions can arise and something can be answered.
```suggestion
```